keps/sig-node/4818-allow-zero-value-for-sleep-action-of-prestop-hook/README.md

```diff
@@ -0,0 +1,564 @@
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The sleep action for the PreStop container lifecycle hook was introduced in KEP 3960. It however doesn’t accept zero as a valid value for the sleep duration seconds. This KEP aims to add support for setting a value of zero with the sleep action of the PreStop hook.
+
+## Motivation
+
+Currently, trying to create a container with a PreStop lifecycle hook with sleep of 0 seconds will throw a validation error like so:
+
+```
+Invalid value: 0: must be greater than 0 and less than terminationGracePeriodSeconds (30)
+```
+
+The Sleep action is implemented with the time package from Go’s standard library. The `time.After()` which is used to implement the sleep permits a zero sleep duration. A negative or a zero sleep duration will cause the function to return immediately and function like a no-op.
+
+The implementation in KEP 3960 supports only non-zero values for the sleep duration. It is semantically correct to support a zero value for this field since time.After() also supports zero and negative durations. Negative values as well as zero have the same effect with time.After(), they both return immediately. We don’t need to support negative values since they have the same effect as setting the duration to zero.
```

Comment:
can you please expand on whether kubelet accepts `0` today and it is only API server validaiton or kubelet also validates the value?

Reply:
Currently, kubelet accepts `0`. No validation is done at the kubelet level. The feature gate for `PodLifecycleSleepAction` is used inside `runSleepHandler` function and it returns nil if the feature is disabled. We can do a similar check at the kubelet level but it doesn't really make sense since a sleep of zero seconds would be similar to returning `nil` right? Since the logic in kubelet already supports zero value (`time.After(0)`), I feel that we don't need the feature gate in kubelet. Also the feature gate check for `PodLifecycleSleepAction` would be removed in v1.32 since its planning to graduate to stable. Please let me know what you think!

[Relevant function in kubelet's lifecyle/handlers.go](https://github.com/AxeZhan/kubernetes/blob/3a96afdfefdf329c637623ae31a61d20dbdb0393/pkg/kubelet/lifecycle/handlers.go#L129-L141)

```go
func (hr *handlerRunner) runSleepHandler(ctx context.Context, seconds int64) error {
	if !utilfeature.DefaultFeatureGate.Enabled(features.PodLifecycleSleepAction) {
		return nil
	}
	c := time.After(time.Duration(seconds) * time.Second)
	select {
	case <-ctx.Done():
		// unexpected termination
		return fmt.Errorf("container terminated before sleep hook finished")
	case <-c:
		return nil
	}
}
```